pimd/pim_mroute.c

```diff
@@ -238,22 +137,21 @@ static int pim_mroute_msg_nocache(int fd, struct interface *ifp,
 	return 0;
 }
 
-static int pim_mroute_msg_wholepkt(int fd, struct interface *ifp,
-				   const char *buf)
+int pim_mroute_msg_wholepkt(int fd, struct interface *ifp, const char *buf)
 {
 	struct pim_interface *pim_ifp;
 	pim_sgaddr sg;
 	struct pim_rpf *rpg;
-	const struct ip *ip_hdr;
+	const ipv_hdr *ip_hdr;
 	struct pim_upstream *up;
 
 	pim_ifp = ifp->info;
 
-	ip_hdr = (const struct ip *)buf;
+	ip_hdr = (const ipv_hdr *)buf;
 
 	memset(&sg, 0, sizeof(sg));
-	sg.src = ip_hdr->ip_src;
```

Comment:
Better to move these lines to `pim[6]_mroute_msg.c` and add a `pim_sgaddr sg` parameter to `pim_mroute_msg_wholepkt()` - no need for `ipv_hdr`, `IPV_SRC` and `IP_DST` then

Reply:
In Wholepkt full packet is required and packet length is also has to be retrieved. So I thought better to keep it this way

Reply:
As already mentioned, payload length is also required. I have kept it this way retrieving all the parameters from packet